Repo Gardening: allow escalating issues to multiple groups

jeherve opened the pull request:
## Proposed changes:

Up until now, once we had sent one Slack notification about an issue, we couldn't send another to a different team. This was problematic given that we attempt to warn about escalated issues to 2 different teams:

- Product Ambassador HEs in the Triage Issues task
- Dev teams in the Update Board task

By splitting things into 2 different labels ("[Status] Priority Review Triggered" for dev teams, "[Status] Escalated to Product Ambassadors" for Product ambassadors), we can now send notifications to two different teams.

It is not the perfect solution since ideally, we'd want to allow notifications per Slack channel instead of per team, but:

- There is currently no better way to warn about an issue already escalated than to add a label.
- We cannot really use the Slack channel ID in the label name, that would be odd.

In the process of making this change, I realized that the `updateBoard` and `triageIssues` tasks had become more and more complex, while incorporating elements common to both tasks. At this point, I think it makes sense to get rid of one task (`updateBoard`) and group everything under a single task, while splitting some of the logic to their own files. I think it's easier to understand that way, and newer contributors should find it easier to get started editing triage tasks. After all, making changes to project boards related to an issue is also triage, no less than making change to the issue's labels.

**This PR consequently ends up making a few extra changes to make the logic (hopefully) a bit easier to grasp.**

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

* N/A

## Does this pull request change what data or activity we track or use?

* No

## Testing instructions:

This is better tested in a fork of this repo, where you will have set up the 2 different tasks already, as well as GitHub secrets to make them active. In that fork, I would also recommend setting up the `labels_team_assignments` input in your fork to allow for the Update Board task to work.

Here is an example where I've tested in my own fork:

https://github.com/jeherve/jetpack/issues/167

jeherve committed:
Repo Gardening: allow escalating issues to multiple groups

Up until now, once we had sent one Slack notification about an issue, we couldn't send another to a different team. This was problematic given that we attempt to warn about escalated issues to 2 different teams:

- Product Ambassador HEs in the Triage Issues task
- Dev teams in the Update Board task

By splitting things into 2 different labels ("[Status] Priority Review Triggered" for dev teams, "[Status] Escalated to Product Ambassadors" for Product ambassadors), we can now send notifications to two different teams.

It is not the perfect solution since ideally, we'd want to allow notifications per Slack channel instead of per team, but:

- There is currently no better way to warn about an issue already escalated than to add a label.
- We cannot really use the Slack channel ID in the label name, that would be odd.

jeherve committed:
Simplify check before we send slack notification

This was added in #30100, but it ends up blocking legitimate messages from being sent.

jeherve committed:
Move checking for labels out of the Slack notification function

This should be handled by each task calling the function, for more flexibility, and a simpler notification function.

jeherve committed:
Remove Update Board task

We'll merge it into the existing triage issues task. It is not very intuitive to have this being two different tasks, when both perform actions that can be described as issue triage. It also makes the codebase more approachable, and will allow us to extract / centralize some of the logic that was used in both tasks.

jeherve committed:
Extract more label detection functions

jeherve committed:
Extract AI labeling into its own file

It should make it a bit easier for folks to contribute.

jeherve committed:
Add new Project board management "sub-task" for triageIssues

jeherve committed:
Document the new actions performed by the Triage issues task

jeherve committed:
Extract method used to find the priority of an issue

jeherve committed:
Bring it all together in the update triageIssues task

jeherve committed:
Fix argument order

jeherve committed:
Add missing argument

jeherve committed:
Remove unnecessary state check (we check earlier) & add logging

jeherve commented on projects/github-actions/repo-gardening/changelog/rm-update-board-task:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: changed
+
+Board Triage: remove updateBoard task. It will now be part of the existing triageIssues task.
```
I've been wondering if I should change this to a major bump. It doesn't create errors in the action if you continue to have the `updateBoard` task set in your workflow file, but that task will just not run, thus breaking your expectations.

jeherve replied:
I've thought about it and switched it to a major bump in fecf1ec153f5a1b079ddb0fa90defb536370d529

jeherve commented on projects/github-actions/repo-gardening/src/utils/labels/has-escalated-label.js:
```diff
@@ -3,34 +3,38 @@ const getLabels = require( './get-labels' );
 	const labels = await getLabels( octokit, owner, repo, number );
-	if (
-		labels.includes( '[Status] Priority Review Triggered' ) ||
-		labels.includes( '[Status] Escalated to Kitkat' )
```
I'm removing this label as part of the refactor, because it isn't used anymore.

jeherve committed:
Remove files that are no longer used

jeherve committed:
Add label when issue is automatically triaged to project board

See pfVjQF-55-p2#comment-119

jeherve committed:
Update label name

See pfVjQF-55-p2#comment-127

jeherve committed:
Switch to major version bump

See https://github.com/Automattic/jetpack/pull/39973#discussion_r1833195889

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -179,93 +70,45 @@ async function triageIssues( payload, octokit ) {
-		// Add priority label to all bugs, if none already exists on the issue.
-		if ( priorityLabels.length === 0 && isBugIssue ) {
-			debug( `triage-issues: Adding [Pri] ${ priority } label to issue #${ number }` );
+		// Add priority label to the issue, if none already existed on the issue.
```
Looks like this code doesn't necessarily do that. It will also waste an API call trying to re-add the already-existing priority label if there's already exactly one.

jeherve replied:
That's a good point. What do you think of 1895c57f2cadf80c8cedabb6b5c4a05d08727cdb to handle that?

anomiex commented on projects/github-actions/repo-gardening/src/tasks/triage-issues/index.js:
```diff
@@ -179,93 +70,45 @@ async function triageIssues( payload, octokit ) {
 			// If we're adding a TBD priority, if we're in the Calypso repo,
 			// send a Slack notification.
-			if ( priority === 'TBD' && full_name === 'Automattic/wp-calypso' ) {
+			if ( inferredPriority === 'TBD' && full_name === 'Automattic/wp-calypso' && qualityChannel ) {
```
This won't ever send the Slack message, because `inferredPriority` will be `[Pri] TBD` now.

jeherve replied:
Good catch! e3b5411b0009e347e60def40322656fa15de0f02

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/has-escalated-label.js:
```diff
@@ -3,34 +3,38 @@ const getLabels = require( './get-labels' );
-		eventLabel.name.match( /^\[Status\] Priority Review Triggered.*$/ )
-	) {
+	// If the issue is being labeled, check if the new label is the escalation label.
+	if ( 'labeled' === action && eventLabel.name && eventLabel.name === escalatedLabel ) {
```
Should we take the opportunity to clean up this similar to 148b28cd?

jeherve replied:
Yes, that would be an opportunity to simplify things a bit. c3b529c79df88c4055574ffc0cf8197a4dd4f387

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/has-triaged-label.js:
```diff
@@ -0,0 +1,27 @@
+ */
+async function hasTriagedLabel( octokit, owner, repo, number, action, eventLabel ) {
+	const labels = await getLabels( octokit, owner, repo, number );
+	if ( 'labeled' === action && eventLabel.name && eventLabel.name === 'Triaged' ) {
```
Should we take the opportunity to clean up this similar to 148b28cd?

jeherve replied:
Yes, that would be an opportunity to simplify things a bit. c3b529c79df88c4055574ffc0cf8197a4dd4f387

anomiex commented on projects/github-actions/repo-gardening/src/utils/labels/needs-third-party-fix.js:
```diff
@@ -0,0 +1,31 @@
+ */
+async function needsThirdPartyFix( octokit, owner, repo, number, action, eventLabel ) {
+	const labels = await getLabels( octokit, owner, repo, number );
+	if ( 'labeled' === action && eventLabel.name ) {
```
Should we take the opportunity to clean up this similar to 148b28cd?

jeherve replied:
Yes, that would be an opportunity to simplify things a bit. c3b529c79df88c4055574ffc0cf8197a4dd4f387

jeherve committed:
Update to new label value format

See https://github.com/Automattic/jetpack/pull/39973#discussion_r1836832119

jeherve committed:
Simplify check for existing labels on issues

See https://github.com/Automattic/jetpack/pull/39973#discussion_r1836851347

This is similar to what was already done in 148b28cd23ca31c9a574cbcecc0d9f9a4e887997

jeherve committed:
Better clarify where issue priority comes from, and use it

Let's not attempt to add a label to an issue when it is already on the issue.

See https://github.com/Automattic/jetpack/pull/39973#discussion_r1836829024

anomiex reviewed (approved):
Let's give it a try. Haven't tested it myself.
